packages/block-editor/src/components/block-list/block.native.js

```diff
@@ -187,18 +269,34 @@ export default compose( [
 		const icon = blockType.icon;
 		const getAccessibilityLabelExtra = blockType.__experimentalGetAccessibilityLabel;
 
-		const selectedBlock = getSelectedBlock();
-		const parentId = getBlockRootClientId( clientId );
-		const parentBlock = getBlock( parentId );
-
-		const isMediaText = selectedBlock && selectedBlock.name === 'core/media-text';
-		const isMediaTextParent = parentBlock && parentBlock.name === 'core/media-text';
+		const parents = getBlockParents( clientId, true );
+		const parentId = parents[ 0 ] || '';
 
 		const rootBlockId = getBlockHierarchyRootClientId( clientId );
 		const rootBlock = getBlock( rootBlockId );
 		const hasRootInnerBlocks = rootBlock.innerBlocks.length !== 0;
 
-		const showFloatingToolbar = isSelected && hasRootInnerBlocks && ! isMediaText && ! isMediaTextParent;
+		const showFloatingToolbar = isSelected && hasRootInnerBlocks;
+
+		const selectedBlockClientId = getSelectedBlockClientId();
+
+		const commonAncestor = getLowestCommonAncestorWithSelectedBlock( clientId );
+		const commonAncestorIndex = parents.indexOf( commonAncestor ) - 1;
+		const firstToSelectId = commonAncestor ? parents[ commonAncestorIndex ] : parents[ parents.length - 1 ];
+
+		const hasChildren = !! getBlockCount( clientId );
+		const hasParent = !! parents[ 0 ];
+		const isParentSelected = selectedBlockClientId && selectedBlockClientId === parents[ 0 ];
+		const isAncestorSelected = selectedBlockClientId && parents.includes( selectedBlockClientId );
+		const isSelectedBlockNested = !! getBlockRootClientId( selectedBlockClientId );
+
+		const isDescendantSelected = selectedBlockClientId && getBlockParents( selectedBlockClientId ).includes( clientId );
+		const isDescendantOfParentSelected = selectedBlockClientId && getBlockParents( selectedBlockClientId ).includes( parentId );
```

Comment:
After reading this comment, it did look a bit related to the `firstToSelect` search, so I think we could have the equivalent (please double check my logic here 😉):

- `isDescendantSelected`: if the selection is one of the descendants, then this must be a parent of the selection, which also means it will be the lowest common ancestor. So `isDescendantSelected = commonAncestor === clientId`
- `isDescendantOfParentSelected`: similar to the previous one, if the parent of this block is an ancestor of the selection, but the selection is not one of this block's descendants, then the lowest common ancestor if this block's parent: `isDescendantOfParentSelected = commonAncestor === parentId`

Reply:
@koke I temporary go with @dratwas suggest.

However I agree with this one `isDescendantSelected = commonAncestor === clientId`

but not sure about this 
`isDescendantOfParentSelected = commonAncestor === parentId` 

Shouldn't we check the common ancestor with parent of selected block ?

Then it will look like this:
`const commonAncestorParent = getLowestCommonAncestorWithSelectedBlock( parentId );`
`const isDescendantOfParentSelected = commonAncestorParent === parentId;`

Reply:
Right, I think that would change the behavior a bit. The `commonAncestor == parentId` would not include the case where `isDescendantSelected == true`. It would be more accurate to say `isDescendantOfSiblingSelected`, which might not be what you had in mind when doing the calculations. I think we can combine both:

```js
const isDescendantSelected = commonAncestor === clientId;
const isDescendantOfParentSelected = isDescendantSelected || commonAncestor === parentId;
```

But at this stage, this feels more like nitpicking, so I'm happy with either approach, we shouldn't block on this one to merge.